aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -71,52 +71,162 @@ pub(crate) fn validate_module_events(
+                use Bytecode::*;
+                match bc {
+                    CallGeneric(index) | PackClosureGeneric(index, ..) => {
+                        let func_instantiation = &module.function_instantiation_at(*index);
+                        let func_handle = module.function_handle_at(func_instantiation.handle);
+                        let module_handle = module.module_handle_at(func_handle.module);
+                        let module_addr = module.address_identifier_at(module_handle.address);
+                        let module_name = module.identifier_at(module_handle.name);
+                        let func_name = module.identifier_at(func_handle.name);
+                        if module_addr != &AccountAddress::ONE
+                            || module_name.as_str() != EVENT_MODULE_NAME
+                            || func_name.as_str() != EVENT_EMIT_FUNCTION_NAME
+                        {
+                            continue;
+                        }
+                        let param = module
+                            .signature_at(func_instantiation.type_parameters)
+                            .0
+                            .first()
+                            .ok_or_else(|| {
+                                metadata_validation_error(
+                                    "Missing parameter for 0x1::event::emit function",
+                                )
+                            })?;
+                        match param {
+                            StructInstantiation(index, _) | Struct(index) => {
+                                let struct_handle = &module.struct_handle_at(*index);
+                                let struct_name = module.identifier_at(struct_handle.name);
+                                if struct_handle.module != module.self_handle_idx() {
+                                    metadata_validation_err(format!("{} passed to 0x1::event::emit function is not defined in the same module", struct_name).as_str())
+                                } else if !event_structs.contains(struct_name.as_str()) {
+                                    metadata_validation_err(format!("Missing #[event] attribute on {}. The #[event] attribute is required for all structs passed into 0x1::event::emit.", struct_name).as_str())
+                                } else {
+                                    Ok(())
+                                }
+                            },
+                            _ => metadata_validation_err(
+                                "Passed in a non-struct parameter into 0x1::event::emit.",
+                            ),
+                        }?;
+                    },
+                    // Note: If a closure is packed, it cannot be 0x1::event::emit, but the lifted
+                    // lambda body may contain the emit function, and so will match the case above.
+                    // For all other instructions, no validation.
+                    PackClosure(_, _)
```

Comment:
nit: `_ => ()` instead of the whole match

Reply:
Actually I did a whole match on purpose so next time a bytecode is added we think if this part of the code needs changes. If there was a match before, this would have been caught during reviews, or at least a thorough reviewer would think about this case and raise any concerns

Reply:
+1 for not having `_ =>`.

Reply:
Can this be a comment?
> This full match has been left on purses to evaluate all ByteCode instructions in case new ones are added.

Reply:
Indeed, added a comment